### PR TITLE
Feat/status updates

### LINKED ISSUE #61: Get Status from provider
Right now each provider updates the pod status itself. This should be done in the Kubelet logic by calling the provider's `status` method. ~Also, status should return an array of container statuses that should be set on the pod as well~

### PR DESCRIPTION
**Pull request by thomastaylor312**
Updating status is now working for wasi. This has all been added as part
of the `wasi_runtime::handle` module, and so should be portable to
wascc when it is ready. It introduces a new `PodHandle` which is a higher
level (and possibly better?) abstraction for interacting with a running Pod.

I also removed the `status` method from the provider interface as it should
be done with handles and not by the kubelet calling the provider.

Closes #61

### CODE REVIEW

**Comment by bacongobbler** [comment]
Looks like there are a few failing CI tests due to calling `.unwrap` on an `&str`. Let me know when this is ready for review

File: crates/kubelet/src/kubelet.rs [hunk 1]
```diff
@@ -37,7 +37,7 @@ pub struct NotImplementedError;
 ///
 /// This is specified by Kubernetes itself.
 #[derive(Clone, Debug, serde::Serialize)]
-pub enum Phase {
+pub(crate) enum Phase {
```
Comment (bacongobbler) [inline crates/kubelet/src/kubelet.rs]: would it be useful for providers to inspect the pod's phase?

File: crates/kubelet/src/server.rs [hunk 2]
```diff
@@ -49,9 +49,9 @@ pub async fn start_webserver<T: 'static + Provider + Send + Sync>(
                             get_container_logs(
                                 &*provider.lock().await,
                                 &req,
-                                namespace.to_string(),
-                                pod.to_string(),
-                                container.to_string(),
+                                (*namespace).to_string(),
```
Comment (bacongobbler) [inline crates/kubelet/src/server.rs]: out of curiosity, what's the reason for this change? How come we have to deref here?

File: crates/wasi-provider/src/lib.rs [hunk 3]
```diff
@@ -69,34 +66,39 @@ impl Provider for WasiProvider {
...
+                PodHandle::new(container_handles, pod, client)?,
+            );
         }
         info!(
             "All containers started for pod {:?}. Updating status",
-            pod.name()
+            pod_name
         );
-        pod.patch_status(client, &Phase::Running).await;
+        // pod.patch_status(client, &Phase::Running).await;
```
Comment (bacongobbler) [inline crates/wasi-provider/src/lib.rs]: should this be removed completely?

File: crates/kubelet/src/kubelet.rs (hunk 1 above)
Comment (thomastaylor312) [reply crates/kubelet/src/kubelet.rs]: Based on how I have it implemented now, there should be no need to check the phase outside of the crate

File: crates/kubelet/src/server.rs (hunk 2 above)
Comment (thomastaylor312) [reply crates/kubelet/src/server.rs]: This is a clippy suggestion. Basically `namespace` is a `&&` which has a slower implementation of to_string

File: crates/wasi-provider/src/lib.rs (hunk 3 above)
Comment (thomastaylor312) [reply crates/wasi-provider/src/lib.rs]: Oops, yes

**Comment by thomastaylor312** [comment]
I'd love to add e2e tests for this! I was actually going to rebase on your PR once you've merged

File: tests/integration_tests.rs [hunk 4]
```diff
@@ -174,12 +174,17 @@ async fn test_wasi_provider() -> Result<(), Box<dyn std::error::Error>> {
...
+    let mut found_running = false;
     while let Some(event) = watcher.try_next().await? {
         match event {
             WatchEvent::Modified(o) => {
                 let phase = o.status.unwrap().phase.unwrap();
                 if phase == "Running" {
+                    found_running = true;
+                }
+                if phase == "Succeeded" && !found_running {
+                    panic!("Reached completed phase before receiving Running phase")
```
Comment (bacongobbler) [inline tests/integration_tests.rs]: could there ever be a case where we see the completed phase before the running phase from the events system, or is the order guaranteed?

Comment (thomastaylor312) [reply tests/integration_tests.rs]: I think eventing order is guaranteed. This was mostly to check that we actually got a running status reported at some point

**Comment by thomastaylor312** [comment]
@bacongobbler Wascc doesn't have any status updating yet, so it is not included in tests. It was just updated to match the new format

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -35,20 +35,23 @@ jobs:
...
+          pathInArchive: just
+      - name: Run e2e tests
+        env:
+          KRUSTLET_NODE_IP: "172.17.0.1"
+          # The default location for the cert is not accessible on build hosts, so do it in the current dir
+          KEY_DIR: "./config"
+          PFX_PATH: "./config/certificate.pfx"
+        run: |
+          ifconfig
+          nslookup `hostname`
```
Comment (rylev) [inline .github/workflows/build.yml]: Can we add a comment to the `ifconfig/nslookup` magic

File: crates/kubelet/src/pod.rs [hunk 6]
```diff
@@ -85,27 +86,88 @@ impl Pod {
...
+            }
+        };
+
+        // This section figures out what the current phase of the pod should be
+        // based on the container statuses
+        let current_statuses = status
+            .container_statuses
+            .into_iter()
+            .map(|s| s.to_kubernetes(name.to_owned()))
+            .collect::<Vec<KubeContainerStatus>>();
```
Comment (rylev) [inline crates/kubelet/src/pod.rs]: Should we collect this into a `HashMap<String, KubeContainerStatus>` where the keys are the `name`s so we can do quick lookups in the filter below?

File: crates/kubelet/src/pod.rs [hunk 7]
```diff
@@ -85,27 +86,88 @@ impl Pod {
...
+            .into_iter()
+            .filter(|s| {
+                current_statuses
+                    .iter()
+                    .find(|&other| other.name == s.name)
+                    .is_none()
+            })
+            .collect::<Vec<KubeContainerStatus>>();
+        container_statuses.extend(current_statuses);
+        let mut num_terminated: usize = 0;
```
Comment (rylev) [inline crates/kubelet/src/pod.rs]: `terminated` can have a connotation of failure. Is `num_succeed` a better name?

File: crates/kubelet/src/server.rs (hunk 2 above)
Comment (rylev) [reply crates/kubelet/src/server.rs]: Out of curiosity, does `to_owned` lead to the same clippy warning?

File: crates/wascc-provider/src/lib.rs [hunk 8]
```diff
@@ -105,10 +105,30 @@ impl Provider for WasccProvider {
                 tokio::task::spawn_blocking(move || wascc_run_http(data, env, &pub_key)).await?;
             match http_result {
                 Ok(_) => {
-                    pod.patch_status(client.clone(), &Phase::Running).await;
+                    pod.patch_status(
+                        client.clone(),
+                        Status {
```
Comment (rylev) [inline crates/wascc-provider/src/lib.rs]: I think all these `patch_status` calls in this PR would be much easier to read if the status arguments were not inlined and instead moved to a local variable.

File: crates/wasi-provider/src/handle.rs [hunk 9]
```diff
@@ -46,25 +52,125 @@ impl<R: AsyncReadExt + AsyncSeekExt + Unpin> RuntimeHandle<R> {
...
+    ) -> anyhow::Result<usize> {
+        let mut handles = self.container_handles.write().await;
+        let handle =
+            handles
+                .get_mut(container_name)
+                .ok_or_else(|| ProviderError::ContainerNotFound {
+                    pod_name: self.pod.name().to_owned(),
+                    container_name: container_name.to_owned(),
+                })?;
+        handle.output(buf).await
```
Comment (rylev) [inline crates/wasi-provider/src/handle.rs]: The way that this is architected, we'll hold on to the write lock while `handle.output` is running. That's ok right?

File: crates/wasi-provider/src/lib.rs [hunk 10]
```diff
@@ -69,34 +66,38 @@ impl Provider for WasiProvider {
         //   - mount any volumes (popen)
         //   - run it to completion
         //   - bail if it errors
-        info!("Starting containers for pod {:?}", pod.name());
+        let pod_name = pod.name().to_owned();
```
Comment (rylev) [inline crates/wasi-provider/src/lib.rs]: Is to_owned needed? Is this to get around borrowing?

File: tests/integration_tests.rs [hunk 11]
```diff
@@ -193,9 +198,20 @@ async fn test_wasi_provider() -> Result<(), Box<dyn std::error::Error>> {
     let mut logs = pods.log_stream("hello-wasi", &LogParams::default()).await?;
 
     while let Some(line) = logs.try_next().await? {
-        assert_eq!("Hello World!\n", String::from_utf8_lossy(&line));
+        assert_eq!("Hello, world!\n", String::from_utf8_lossy(&line));
     }
 
+    let pod = pods.get("hello-wasi").await?;
+
+    let state = pod.status.unwrap().container_statuses.unwrap()[0]
```
Comment (rylev) [inline tests/integration_tests.rs]: This is super gnarly. I know that this is a test but can we clean this up?

File: crates/wasi-provider/src/handle.rs [hunk 12]
```diff
@@ -46,25 +52,125 @@ impl<R: AsyncReadExt + AsyncSeekExt + Unpin> RuntimeHandle<R> {
...
-    // need to expose it, we can do that later
     pub(crate) async fn wait(&mut self) -> anyhow::Result<()> {
         (&mut self.handle).await.unwrap()
     }
 }
+
+/// PodHandle is the top level handle into managing a pod. It manages updating
+/// statuses for the containers in the pod and can be used to stop the pod and
+/// access logs
+pub struct PodHandle<R: AsyncReadExt + AsyncSeekExt + Unpin> {
```
Comment (rylev) [inline crates/wasi-provider/src/handle.rs]: I don't think you can just use `AsyncRead + AsyncSeek`

File: tests/integration_tests.rs (hunk 11 above)
Comment (thomastaylor312) [reply tests/integration_tests.rs]: Have a way of doing it that doesn't involve another function? I tried a couple different things and none of them worked

File: crates/wasi-provider/src/handle.rs (hunk 9 above)
Comment (thomastaylor312) [reply crates/wasi-provider/src/handle.rs]: Yes, because we need the mutable lock on that handle until we are done

File: crates/wasi-provider/src/lib.rs (hunk 10 above)
Comment (thomastaylor312) [reply crates/wasi-provider/src/lib.rs]: Yes. This was because we need to move the pod into the new `PodHandle` further down

File: .github/workflows/build.yml (hunk 5 above)
Comment (thomastaylor312) [reply .github/workflows/build.yml]: This was just debug material while I am getting the e2e tests working

File: crates/kubelet/src/pod.rs (hunk 7 above)
Comment (thomastaylor312) [reply crates/kubelet/src/pod.rs]: Yes it is. Will change

File: crates/wasi-provider/src/handle.rs (hunk 12 above)
Comment (thomastaylor312) [reply crates/wasi-provider/src/handle.rs]: I think I need Ext here as well. As in the previous PR, I don't think you can use the methods I do without those traits in scope

File: crates/wascc-provider/src/lib.rs (hunk 8 above)
Comment (thomastaylor312) [reply crates/wascc-provider/src/lib.rs]: This better?

**Comment by thomastaylor312** [comment]
@rylev Going to merge this for now. Completely open to a follow up PR or an issue for handling any other things you think we should fix
